docs/reference/docs/data-replication.asciidoc

```diff
@@ -0,0 +1,136 @@
+for validating the operation and forwarding it to the other replicas. Since replicas can be off-line, the primary
+is not required to replicate to all replicas. Instead, Elasticsearch maintains a list of shard copies that should
+receive the operation. This list is called the “in-sync copies” and is maintained by the master node. As the name implies,
+these are the set of "good" shard copies that are guaranteed to have processed all of the index and delete operations that
+have been acknowledged to the user. The primary is responsible for maintaining this invariant and thus has to replicate all
+operations to each copy in this set.
+
+The primary shard follows this basic flow:
+
+. Validate incoming operation and reject it if structurally invalid (Example: have an object field where a number is expected)
+. Execute the operation locally - i.e., indexing or deleting the relevant document. This will also validate the content of fields
+   and reject if needed (Example: a keyword value is too long for indexing in Lucene).
+. Forward the operation to each replica in the current in-sync copies set. If there are multiple replicas, this is done concurrently.
+. Once all replicas have successfully performed the operation and responded to the primary, the primary acknowledges the successful
+   completion of the request to the client.
+
+==== Failure handling
+
+Many things can go wrong during indexing - disks can get corrupted, nodes can be disconnected from each other or some
+configuration mistake could cause an operation to fail on a replica despite it being successful on the primary. These
+are infrequent but the primary has to respond to them.
+
+In the case that the primary itself fails, a message will be sent to the master about it. The indexing operation will
+wait (up to 1 minute, by [default](link to doc)) for the master to promote one of the replicas to be a new primary.
+The operation will then be forwarded to the new primary for processing. Note that the master also monitors the health
+of the nodes and may decide to pro-actively demote a primary. This typically happens when the node holding the primary
+ is isolated from the cluster by a networking issue. More on this later.
+
+Once the operation has been  successfully performed on the primary, the primary has to deal with potential failures
+when executing it on the replica shards. This may be caused by an actual failure on the replica or due to a network
+issue preventing the operation from reaching the replica (or preventing the replica from responding). All of these
+share the same end result - a replica which is part of the in-sync replica set misses an operation that is about to
+be acknowledged. In order to avoid violating the invariant, the primary sends a message to  the master requesting
+that the problematic shard be removed from the in-sync replica set. Only once this has successfully been done, does
```

Comment:
here (and above) you use "in-sync replica set" while in other places you use "in-sync copies".  I think I prefer "in-sync replica set"